File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
def dfs(graph, cur, apt, count):
  dx = [1, -1, 0, 0]
  dy = [0, 0, 1, -1]
  graph[cur[0]][cur[1]] = apt
  count[apt] += 1

  for i in range(4):
    new_x = cur[0] + dx[i] 
    new_y = cur[1] + dy[i]
    if 0 <= new_x < len(graph) and 0 <= new_y < len(graph) and graph[new_x][new_y] == 1:
      dfs(graph, (new_x, new_y), apt, count)


def solution():
  n = int(input())
  graph = []
  for _ in range(n):
    graph.append(list(map(int, input())))
  apt = 2
  count = [ 0 for _ in range(625)]
  for x in range(n):
    for y in range(n):
      if graph[x][y] == 1:
        dfs(graph, (x, y), apt, count)
        apt += 1
  filter_count = list(filter(lambda x: x != 0, count))
  filter_count.sort()
  print(apt - 2)
  for cnt in filter_count:
    print(cnt)
 
solution()
  

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const N = Number(input[0]);

const graph = [];
const answer = [];

let max = -1;
let depth = 0;

for (let i = 1; i <= N; i++) {
    graph.push(input[i].split('').map(Number));
}

const DFS = (curX, curY) => {
    depth++;
    max = Math.max(max, depth);
    graph[curX][curY] = -1;
    const dx = [1, -1, 0, 0];
    const dy = [0, 0, 1, -1];

    for (let i = 0; i < 4; i++) {
        const newX = curX + dx[i];
        const newY = curY + dy[i];
        if (newX >= 0 && newY >= 0 && newX < N && newY < N && graph[newX][newY] === 1) {
            DFS(newX, newY);
        }
    }
}

for (let i = 0; i < graph.length; i++) {
    for (let j = 0; j < graph[i].length; j++) {
        if (graph[i][j] === 1) {
            max = -1;
            depth = 0;
            DFS(i, j);
            answer.push(max);
        }
    }
}

answer.sort((a, b) => a - b);

let stringAnswer = answer.length.toString() + '\n';

for (let i = 0; i < answer.length; i++) {
    stringAnswer = stringAnswer + answer[i] + '\n';
}

console.log(stringAnswer.trim());